### ISSUE TITLE
Question: How to modify settings without altering repo files ?

### ISSUE THREAD

**Issue by Halfwalker** [sorin-ionescu/prezto#1832]
Very simple and possibly dumb question ...

I would like to change some settings for zprezto (theme, modules to load, etc.).  I would rather NOT modify the _.zprezto/runcoms/zpreztorc_ file directly - makes updating a pain with git.

I've tried adding **zstyle ':prezto:module:prompt' theme 'powerlevel10k'** to my _.zshrc_ both before and after the **source "${ZDOTDIR:-$HOME}/.zprezto/init.zsh"** but neither seems to make a difference.
Does something have to be un-defined first, or similar ?

Desired end-result - ability to **git pull** without messing with my own modifications.

**Comment by jeffcox** [comment]
What you're asking for isn't really possible, but if you make only minor changes to rc files you can simply `git stash && git pull && git submodule update --init --recursive && git stash apply`

This will eventually fail as the project moves forward, which is why the recommendation is to maintain your own fork.

**Comment by Halfwalker** [comment]
Yeah, thought so.  Darn.
Maybe think of using a drop-in directory type thing - one can drop in snippets of config.  Then zprezto loads those first, and just doesn't reset any that are already defined ?

**Comment by belak** [comment]
Personally, as we rarely update the runcoms, it should be fine to copy those rather than symlink them. I do something similar to avoid maintaining my own fork for my personal config.

**Comment by srijanshetty** [comment]
I would agree with @belak on this one (no surprises there). My personal fork is a token fork. I symlink .zpreztorc and maintain my personal settings in a .zpreto-contrib module. This workflow has worked really well for over half a decade.

Except `.zpreztorc`, all other runcoms haven't been modified in the last three years, which add to the maturity of those runcoms. You'll see rare updates to `.zpreztorc` but you can skim through the changelog once in a while to understand if you want to add those changes to your personal runcom.

**Comment by srijanshetty** [comment]
Should we close this issue @Halfwalker?

**Comment by Halfwalker** [comment]
I use ansible for setting up zsh and zprezto - nice way to ensure things are set up exactly how I like.  My .zshrc is fairly small, basic settings, include my _~/ZSH/zshrc.zprezto_ (all of which _could_ be in .zshrc) and do some zplug things.

The _~/ZSH/zshrc.zprezto_ sets all the POWERLEVEL9k variables, then includes *source "${ZDOTDIR:-$HOME}/.zprezto/init.zsh"*, and finally sets some additional zstyle things.

The ansible role does a couple of modifications to _runcoms/zpreztorc_ and all is well in the world.  The hassle (if one can really call it a hassle) is every time the role runs (I run it a lot as I modify things) it always has to git pull and re-modify that file.  Not really a big deal, but ... unclean.

_runcoms/zpreztorc_ only has 3 settings being made - the rest are commented out and optional.  Perhaps it might looks like this
```
~ 119 zstyle -g theme_test ':prezto:module:prompt' theme                              
+ 120 if [ -z ${theme_test} ] ; then                                                  
+ 121     zstyle ':prezto:module:prompt' theme 'sorin'                                
+ 122 fi 
```
So it's only setting the theme if it's not already set.

This works.  I can set the theme in my own _~/ZSH/zshrc.zprezto_ right after the *source "${ZDOTDIR:-$HOME}/.zprezto/init.zsh"* and the proper powerlevel10k theme is loaded and everything works :)

Basically, before making any setting, just check if it already exists.  If not, set it, otherwise continue.

**Comment by Halfwalker** [comment]
Checking on the logic in _~/.zprezto/init.sh_ it looks like it loads all the modules (pmodules) at the end, **after** sourcing _~/.zprezto/runcoms/zpreztorc_.  I see one way around loading a custom set of modules without modifying the stock file.

Set the modules **before** calling **source "${ZDOTDIR:-$HOME}/.zprezto/init.zsh"**
```
 61 zstyle ':prezto:load' pmodule \                                                 
 62   'environment' 'terminal' 'editor' 'history' 'directory' 'spectrum' \          
 63   'utility' 'completion' 'git' 'tmux' 'syntax-highlighting' \                   
 64   'history-substring-search' 'autosuggestions' 'prompt'                                                                                                     
 65                                                                                 
 66 if [ -e ~/.zprezto/runcoms/zshrc ] ; then                                       
 67     source "${ZDOTDIR:-$HOME}/.zprezto/init.zsh"                                
 68 fi
```  
That works fine.

We can't manually load (source) the modules, because several need to be loaded **before** the _prompt_ module.  No matter, doing the zstyle here works fine, just needs the **if [ -z ....]** check done in _runcoms/zpreztorc_

Hopefully you'll consider adding those checks in :)

**Comment by belak** [comment]
It loads all the modules at the end after sourcing ~/.zpreztorc - the zpreztorc file is meant to be a user configuration file, but we provide a default one if you symlink or copy it to the proper location. We also are fairly careful to not change behavior if something isn't set that doesn't already exist.

My recommendation would be to either customize zpreztorc (and not commit it to the repo - you probably want it in some sort of dotfiles repo for yourself) or customize the zshrc and not include a zpreztorc.

**Comment by Halfwalker** [comment]
Any chance of just adding the checks to _runcoms/zpreztorc_ ?  It wouldn't alter the current behavior.
```
diff --git a/runcoms/zpreztorc b/runcoms/zpreztorc
index 9a3c19a..16d83b8 100644
--- a/runcoms/zpreztorc
+++ b/runcoms/zpreztorc
@@ -29,7 +29,9 @@ zstyle ':prezto:*:*' color 'yes'

 # Set the Prezto modules to load (browse modules).
 # The order matters.
-zstyle ':prezto:load' pmodule \
+zstyle -g prezto_test ':prezto:load' pmodule
+if [ -z ${prezto_test} ] ; then
+  zstyle ':prezto:load' pmodule \
   'environment' \
   'terminal' \
   'editor' \
@@ -39,6 +41,8 @@ zstyle ':prezto:load' pmodule \
   'utility' \
   'completion' \
   'prompt'
+fi
+unset prezto_test

 #
 # Autosuggestions
@@ -60,7 +64,11 @@ zstyle ':prezto:load' pmodule \
 #

 # Set the key mapping style to 'emacs' or 'vi'.
-zstyle ':prezto:module:editor' key-bindings 'emacs'
+zstyle -g prezto_test ':prezto:module:editor' key-bindings
+if [ -z ${prezto_test} ] ; then
+  zstyle ':prezto:module:editor' key-bindings 'emacs'
+fi
+unset prezto_test

 # Auto convert .... to ../..
 # zstyle ':prezto:module:editor' dot-expansion 'yes'
@@ -116,7 +124,11 @@ zstyle ':prezto:module:editor' key-bindings 'emacs'
 # Set the prompt theme to load.
 # Setting it to 'random' loads a random theme.
 # Auto set to 'off' on dumb terminals.
-zstyle ':prezto:module:prompt' theme 'sorin'
+zstyle -g prezto_test ':prezto:module:prompt' theme
+if [ -z ${prezto_test} ] ; then
+  zstyle ':prezto:module:prompt' theme 'sorin'
+fi
+unset prezto_test

 # Set the working directory prompt display length.
 # By default, it is set to 'short'. Set it to 'long' (without '~' expansion)
```
If not I'll go with a customized _zpreztorc_

**Comment by Halfwalker** [comment]
FYI - I'm using a customized _zpreztorc_ now, so I guess you can close this.  I still think it would be nice to have those checks in there, so one can use the default provided _zpreztorc_ - or maybe rename it to _zpreztorc.example_
